A86-xds-http-connect.md

```diff
@@ -0,0 +1,107 @@
+A86: xDS-Based HTTP CONNECT
+----
+* Author(s): @markdroth
+* Approver: @ejona86, @dfawley
+* Status: {Draft, In Review, Ready for Implementation, Implemented}
+* Implemented in: <language, ...>
+* Last updated: 2024-09-25
+* Discussion at: https://groups.google.com/g/grpc-io/c/WiqQ7h003fE
+
+## Abstract
+
+This document provides a design for gRPC to support configuring HTTP
+CONNECT functionality via xDS.
+
+## Background
+
+gRPC currently supports HTTP CONNECT proxies, as described in [gRFC A1].
+gRPC also supports obtaining configuration via xDS, as originally
+described in [gRFC A27].  This document describes new functionality to
+allow configuring use of an HTTP CONNECT proxy via xDS.
+
+### Related Proposals: 
+* [gRFC A1: HTTP CONNECT Proxy Support][gRFC A1]
+* [gRFC A27: xDS-Based Global Load Balancing][gRFC A27]
+* [gRFC A29: xDS-Based Security for gRPC Clients and Servers][gRFC A29]
+* [gRFC A83: A83: xDS GCP Authentication Filter][gRFC A83]
+
+[gRFC A1]: A1-http-connect-proxy-support.md
+[gRFC A27]: A27-xds-global-load-balancing.md
+[gRFC A29]: A29-xds-tls-security.md
+[gRFC A83]: A83-xds-gcp-authn-filter.md
+
+## Proposal
+
+The xDS configuration for HTTP CONNECT is done via the
+[`Http11ProxyUpstreamTransport`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/http_11_proxy/v3/upstream_http_11_connect.proto#L36)
+transport socket wrapper.  This wrapper is sent in the CDS [`transport_socket`
+field](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/config/cluster/v3/cluster.proto#L1099),
+as a wrapper around whatever transport socket is actually used.
+
+Note that the only transport socket implementation that gRPC currently
+supports is `UpstreamTlsContext`, which is used to configure TLS
```

Comment:
To be clear, this means you can't use this feature with plaintext, right? `Http11ProxyUpstreamTransport.transport_socket` is a required field, so if you use `Http11ProxyUpstreamTransport` then you must also be using `UpstreamTlsContext` (until we add more).

Reply:
Good catch!  I think we *do* want this to work with plaintext.  I hadn't noticed that `Http11ProxyUpstreamTransport.transport_socket` is a required field, so I was assuming it could be left empty, just as the `Cluster.transport_socket` field is.

@tonya11en, is there a reason that [`Http11ProxyUpstreamTransport.transport_socket`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/http_11_proxy/v3/upstream_http_11_connect.proto#L38) is a required field?  Given that the top-level [`Cluster.transport_socket`](https://github.com/envoyproxy/envoy/blob/1a153166a6d1e9336ee8982d1a00ba98655c9d39/api/envoy/config/cluster/v3/cluster.proto#L1099) field is optional and defaults to plaintext, why doesn't the one inside of `Http11ProxyUpstreamTransport` work the same way?  Is that something we can just fix in xDS and Envoy?

The alternative is for gRPC to add support for the [`RawBuffer`](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/raw_buffer/v3/raw_buffer.proto#L18) transport socket, which I believe is used to explicitly configure plaintext in xDS.  But I'm honestly not sure why that proto exists in the first place; it seems like somewhat pointless bloat on the wire, given that we default to plaintext anyway if the top-level field is unset.

Reply:
It's was required because [RawBuffer](https://github.com/envoyproxy/envoy/blob/d6120f3c769e70c988ddcc5c7e9cbc2737b5f63c/api/envoy/extensions/transport_sockets/raw_buffer/v3/raw_buffer.proto#L18) exists, but I see no reason why it _has_ to be. We could just have it default to the raw buffer transport, make the field optional, and call it a day if that works for you guys? I'm assuming this is okay, since it's not considered stable yet.

Reply:
Yeah, that sounds like the best approach to me.  Can you please go ahead and make that change in the xDS protos and associated docs?  Thanks!

Reply:
This change is being made in https://github.com/envoyproxy/envoy/pull/36414.

I've updated the wording here to attempt to make the behavior clear.